pkg/ccl/sqlproxyccl/balancer/balancer_test.go

```diff
@@ -37,6 +47,166 @@ func TestBalancer(t *testing.T) {
 	})
 }
 
+func TestRebalancer_processQueue(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+
+	b, err := NewBalancer(ctx, stopper, MaxConcurrentRebalances(2))
+	require.NoError(t, err)
+
+	// Use a custom time source for testing.
+	t0 := time.Date(2000, time.January, 1, 0, 0, 0, 0, time.UTC)
+	timeSource := timeutil.NewManualTime(t0)
+
+	eventCh := make(chan struct{})
+	b.testingKnobs.afterProcessQueueItem = func() {
+		eventCh <- struct{}{}
+	}
+
+	t.Run("retries_up_to_maxTransferAttempts", func(t *testing.T) {
+		count := 0
+		conn := &testBalancerConnHandle{
+			onTransferConnection: func() error {
+				count++
+				return errors.New("cannot transfer")
+			},
+		}
+		req := &rebalanceRequest{
+			createdAt: timeSource.Now(),
+			conn:      conn,
+			dst:       "foo",
+		}
+		b.queue.enqueue(req)
+
+		// Wait until the item has been processed.
+		<-eventCh
+
+		// Ensure that we only retried up to 3 times.
+		require.Equal(t, 3, count)
+	})
+
+	t.Run("conn_was_transferred_by_other", func(t *testing.T) {
+		count := 0
+		conn := &testBalancerConnHandle{}
+		conn.onTransferConnection = func() error {
+			count++
+			// Simulate that connection was transferred by someone else.
+			conn.remoteAddr = "foo"
+			return errors.New("cannot transfer")
+		}
+		req := &rebalanceRequest{
+			createdAt: timeSource.Now(),
+			conn:      conn,
+			dst:       "foo",
+		}
+		b.queue.enqueue(req)
+
+		// Wait until the item has been processed.
+		<-eventCh
+
+		// We should only retry once.
+		require.Equal(t, 1, count)
+	})
+
+	t.Run("conn_was_transferred", func(t *testing.T) {
+		count := 0
+		conn := &testBalancerConnHandle{}
+		conn.onTransferConnection = func() error {
+			count++
+			conn.remoteAddr = "foo"
+			return nil
+		}
+		req := &rebalanceRequest{
+			createdAt: timeSource.Now(),
+			conn:      conn,
+			dst:       "foo",
+		}
+		b.queue.enqueue(req)
+
+		// Wait until the item has been processed.
+		<-eventCh
+
+		// We should only retry once.
+		require.Equal(t, 1, count)
+	})
+
+	t.Run("conn_was_closed", func(t *testing.T) {
+		count := 0
+		conn := &testBalancerConnHandle{}
+		conn.onTransferConnection = func() error {
+			count++
+			return context.Canceled
+		}
+		req := &rebalanceRequest{
+			createdAt: timeSource.Now(),
+			conn:      conn,
+			dst:       "foo",
+		}
+		b.queue.enqueue(req)
+
+		// Wait until the item has been processed.
+		<-eventCh
+
+		// We should only retry once.
+		require.Equal(t, 1, count)
+	})
+
+	t.Run("limit_concurrent_rebalances", func(t *testing.T) {
+		var wg sync.WaitGroup
+		var mu struct {
+			syncutil.Mutex
+			count int
+		}
+		b.testingKnobs.beforeProcessQueueItem = func() {
```

Comment:
nit: Instead of adding these two hooks you can increment count before <-waitCh in the onTransferConnection and decrement it after <-waitCh.